src/hasSubObject.ts

```diff
@@ -0,0 +1,58 @@
+import { equals } from "./equals";
+import { purry } from "./purry";
+
+/**
+ * Checks if `subObject` is a sub-object of `object`, which means for every
+ * property and value in `subObject`, there's the same property in `object`
+ * with an equal value. Equality is checked with `equals`.
+ * @param object the object to test
+ * @param subObject the sub-object to test against
+ * @signature
+ *    R.hasSubObject(object, subObject)
+ * @example
+ *    R.hasSubObject({ a: 1, b: 2, c: 3 }, { a: 1, c: 3 }) //=> true
+ *    R.hasSubObject({ a: 1, b: 2, c: 3 }, { b: 4 }) //=> false
+ *    R.hasSubObject({ a: 1, b: 2, c: 3 }, {}) //=> true
+ * @dataFirst
+ * @category Object
+ */
+export function hasSubObject<S extends object, T extends S = S>(
+  object: T,
+  subObject: S,
+): object is S & T;
```

Comment:
The typing here is wrong; it won't be able to narrow T properly, and the "roles" should be reversed. I couldn't get the type to work without using Simplify, probably because typescript doesn't try to "flatten" the object, but it works with it. Finally, just a nit of mine, but avoid using `object` as a name, as it's also a type, which could lead to confusion and even bugs.

```suggestion
export function hasSubObject<T, S extends Partial<T>>(
  data: T,
  subObject: S,
): data is Simplify<S & T>;
```

This is the test I used to play around with the type:
```ts
  test("narrowing", () => {
    const obj = {} as { a?: string; b?: number };

    if (hasSubObject(obj, { a: "a" })) {
      expectTypeOf(obj).toEqualTypeOf<{ a: string; b?: number }>();
    }

    if (hasSubObject(obj, { a: "a" } as const)) {
      expectTypeOf(obj).toEqualTypeOf<{ a: "a"; b?: number }>();
    }

    // @ts-expect-error [ts2353]
    hasSubObject(obj, { c: true });
  });
```

Reply:
if we use this type, then how would we type the data-last version? 
```ts
export function hasSubObject<T, S extends Partial<T>>(
  subObject: S,
): (data: T) => data is Simplify<S & T>;

// no error:
hasSubObject({ a: 1 })({ b: 2 });
```

should the data-last type remain the same?

also, i think the partial here should be a deep partial, if we want

```ts
hasSubObject({ a: { b: 1, c: 2 } }, { a: { b: 1 } });
```

to be ok